Add "Overrides" support to test harness

ellismg opened the pull request:
This is just the overrides part of #2212.  Landing this as is lets us unblock testing in home and examples at head.

#2212 is going to take a little longer to land because the move to install the package under test using `yarn install` is causing slowdown in CI and the problem seems to be much worse on Node 11 (presumably due to the gRPC issues there where we end up having to build multiple times).

So landing this other part first so we can at least start getting a fair amount of coverage.

lukehoban commented on pkg/testing/integration/program.go:
```diff
@@ -1152,6 +1194,37 @@ func (pt *programTester) prepareNodeJSProject(projinfo *engine.Projinfo) error {
 
+// readPackageJSON unmarshals the package.json file located in pathToPackage. It ensures the returned object
+// has "dependencies", "devDependencies", and "peerDependencies" sections.
+func readPackageJSON(pathToPackage string) (map[string]interface{}, error) {
```
The second sentence in the  comment doesn't appear to be correct (anymore?).

ellismg replied:
Yes, fixed.

lukehoban commented on pkg/testing/integration/util.go:
```diff
@@ -30,6 +30,26 @@ import (
 )
 
+// DecodeMapString takes a string of the form key1=value1:key2=value2 and returns a go map.
+func DecodeMapString(val string) (map[string]string, error) {
```
This doesn't appear to be used?

ellismg replied:
It's a utility function I'll want to use in some repositories to compute the set of overrides to add based on information from NPM.  I wanted to centralize that logic, so I put it here.

lukehoban commented on examples/examples_test.go:
```diff
@@ -81,17 +78,6 @@ func TestExamples(t *testing.T) {
 			Dependencies: []string{"@pulumi/pulumi"},
 		},
-		{
-			Dir:          path.Join(cwd, "compat/v0.10.0/minimal"),
```
Why did we drop this test?  Can't we replace with a version of "minimal" that uses the new "overrides"?

ellismg replied:
The test only works on 6.10.X because that's the only version of node 0.10.0 supported. It was passing everywhere because it incorrectly linked in the version of the just built `@pulumi/pulumi` package.

I changed things up such that we will run the test on 6.10.X, but on other nodes, we just skip it.

ellismg commented:
I believe that I have addressed all of your feedback, @lukehoban.  I'm going to merge this as is because it is required to turn on the automated runs in home and examples.  If you have any more changes you'd like, I'm happy to make them in a follow up PR.
